### PR TITLE
Remove Linked list

### PR DESCRIPTION
**Pull request by kalaiyarasiganeshalingam**
## Purpose
This pr removes the Linked list and improves the existing performance too.

Resolves: [#761](https://github.com/ballerina-platform/ballerina-standard-library/issues/761)

**The performance test result** 

**ballerina-swan-lake-alpha1**

**New implementation**:
```
____________________Test performance for LRU______________________
Testing performance with the cache capacity: 10
Cache put time is 2311922672 (ns) and hit rate is 1000000
Worker time: 716176364 (ns); Hit rate: 1000000; Started at: 0
Worker time: 720215314 (ns); Hit rate: 1000000; Started at: 0
Worker time: 721767772 (ns); Hit rate: 1000000; Started at: 0
Worker time: 722742935 (ns); Hit rate: 1000000; Started at: 0
Total time: 5 (s)
Testing performance with the cache capacity: 100
Cache put time is 1796288815 (ns) and hit rate is 1000000
Worker time: 736050481 (ns); Hit rate: 1000000; Started at: 0
Worker time: 737158639 (ns); Hit rate: 1000000; Started at: 0
Worker time: 756019281 (ns); Hit rate: 1000000; Started at: 0
Worker time: 743187750 (ns); Hit rate: 1000000; Started at: 0
Total time: 4 (s)
Testing performance with the cache capacity: 1000
Cache put time is 1806738120 (ns) and hit rate is 1000000
Worker time: 760183132 (ns); Hit rate: 1000000; Started at: 0
Worker time: 761146818 (ns); Hit rate: 1000000; Started at: 0
Worker time: 767975347 (ns); Hit rate: 1000000; Started at: 0
Worker time: 774904365 (ns); Hit rate: 1000000; Started at: 0
Total time: 4 (s)
Testing performance with the cache capacity: 10000
Cache put time is 2109293194 (ns) and hit rate is 1000000
Worker time: 791466072 (ns); Hit rate: 1000000; Started at: 0
Worker time: 794507231 (ns); Hit rate: 1000000; Started at: 0
Worker time: 785123034 (ns); Hit rate: 1000000; Started at: 0
Worker time: 808625927 (ns); Hit rate: 1000000; Started at: 0
Total time: 5 (s)
Testing performance with the cache capacity: 100000
Cache put time is 2614991199 (ns) and hit rate is 1000000
Worker time: 970266336 (ns); Hit rate: 1000000; Started at: 0
Worker time: 903942759 (ns); Hit rate: 1000000; Started at: 0
Worker time: 1070018474 (ns); Hit rate: 1000000; Started at: 0
Worker time: 921128910 (ns); Hit rate: 1000000; Started at: 0
Total time: 6 (s)
```

**Existing LRU policy:**
```
Testing performance with the cache capacity: 10
Cache put time is 3722438807 (ns) and hit rate is 1000000
Worker time: 1565222207 (ns); Hit rate: 1000000; Started at: 0
Worker time: 1399268762 (ns); Hit rate: 1000000; Started at: 0
Worker time: 1394150429 (ns); Hit rate: 1000000; Started at: 0
Worker time: 1387722024 (ns); Hit rate: 1000000; Started at: 0
Total time: 9 (s)
Testing performance with the cache capacity: 100
Cache put time is 3714018324 (ns) and hit rate is 1000000
Worker time: 1528798889 (ns); Hit rate: 1000000; Started at: 0
Worker time: 1519738052 (ns); Hit rate: 1000000; Started at: 0
Worker time: 1517876383 (ns); Hit rate: 1000000; Started at: 0
Worker time: 1528530570 (ns); Hit rate: 1000000; Started at: 0
Total time: 9 (s)
Testing performance with the cache capacity: 1000
Cache put time is 3236667585 (ns) and hit rate is 1000000
Worker time: 1575415759 (ns); Hit rate: 1000000; Started at: 0
Worker time: 1554968022 (ns); Hit rate: 1000000; Started at: 0
Worker time: 1549550258 (ns); Hit rate: 1000000; Started at: 0
Worker time: 1559706780 (ns); Hit rate: 1000000; Started at: 0
Total time: 9 (s)
Testing performance with the cache capacity: 10000
Cache put time is 4587288149 (ns) and hit rate is 1000000
Worker time: 1635438141 (ns); Hit rate: 1000000; Started at: 0
Worker time: 1594387592 (ns); Hit rate: 1000000; Started at: 0
Worker time: 1581636965 (ns); Hit rate: 1000000; Started at: 0
Worker time: 1585630380 (ns); Hit rate: 1000000; Started at: 0
Total time: 10 (s)
Testing performance with the cache capacity: 100000
Cache put time is 3829329858 (ns) and hit rate is 1000000
Worker time: 1842044371 (ns); Hit rate: 1000000; Started at: 0
Worker time: 1662423517 (ns); Hit rate: 1000000; Started at: 0
Worker time: 1665962940 (ns); Hit rate: 1000000; Started at: 0
Worker time: 1629019365 (ns); Hit rate: 1000000; Started at: 0
Total time: 10 (s)
```
**ballerina-swan-lake-beta1**

**New implementation**
```
Testing performance with the cache capacity: 10
Cache put time is 7050892000.000000 (ns) and hit rate is 1000000
Worker time: 3468529000.000000 (ns); Hit rate: 1000000; Started at: 0
Worker time: 3380534000.000000 (ns); Hit rate: 1000000; Started at: 0
Worker time: 3333509000.000000 (ns); Hit rate: 1000000; Started at: 0
Worker time: 3326581000.000000 (ns); Hit rate: 1000000; Started at: 0
Total time: 20.565301 (s)
Testing performance with the cache capacity: 100
Cache put time is 6984036000.000000 (ns) and hit rate is 1000000
Worker time: 3422958000.000000 (ns); Hit rate: 1000000; Started at: 0
Worker time: 3401562000.000000 (ns); Hit rate: 1000000; Started at: 0
Worker time: 3397564000.000000 (ns); Hit rate: 1000000; Started at: 0
Worker time: 3394995000.000000 (ns); Hit rate: 1000000; Started at: 0
Total time: 20.603081 (s)
Testing performance with the cache capacity: 1000
Cache put time is 6155999000.000000 (ns) and hit rate is 1000000
Worker time: 3423339000.000000 (ns); Hit rate: 1000000; Started at: 0
Worker time: 3407754000.000000 (ns); Hit rate: 1000000; Started at: 0
Worker time: 3434907000.000000 (ns); Hit rate: 1000000; Started at: 0
Worker time: 3430001000.000000 (ns); Hit rate: 1000000; Started at: 0
Total time: 19.853862 (s)
Testing performance with the cache capacity: 10000
Cache put time is 7113276000.000000 (ns) and hit rate is 1000000
Worker time: 3483067000.000000 (ns); Hit rate: 1000000; Started at: 0
Worker time: 3437827000.000000 (ns); Hit rate: 1000000; Started at: 0
Worker time: 3455444000.000000 (ns); Hit rate: 1000000; Started at: 0
Worker time: 5613161000.000000 (ns); Hit rate: 1000000; Started at: 0
Total time: 23.104992 (s)
Testing performance with the cache capacity: 100000
Cache put time is 8576908000.000000 (ns) and hit rate is 1000000
Worker time: 3783567000.000000 (ns); Hit rate: 1000000; Started at: 0
Worker time: 3679730000.00000 (ns); Hit rate: 1000000; Started at: 0
Worker time: 3674837000.000000 (ns); Hit rate: 1000000; Started at: 0
Worker time: 3812971000.000000 (ns); Hit rate: 1000000; Started at: 0
Total time: 23.530033 (s)
```
**Existing LRU policy**
```
Testing performance with the cache capacity: 10
Cache put time is 11030585000.000000 (ns) and hit rate is 1000000
Worker time: 4915418000.000000 (ns); Hit rate: 1000000; Started at: 0
Worker time: 4852545000.000000 (ns); Hit rate: 1000000; Started at: 0
Worker time: 4870276000.000000 (ns); Hit rate: 1000000; Started at: 0
Worker time: 4590353000.000000 (ns); Hit rate: 1000000; Started at: 0
Total time: 30.263837 (s)
Testing performance with the cache capacity: 100
Cache put time is 10311207000.000000 (ns) and hit rate is 1000000
Worker time: 4604460000.00000 (ns); Hit rate: 1000000; Started at: 0
Worker time: 4566696000.000000 (ns); Hit rate: 1000000; Started at: 0
Worker time: 4575931000.000000 (ns); Hit rate: 1000000; Started at: 0
Worker time: 4604802000.000000 (ns); Hit rate: 1000000; Started at: 0
Total time: 28.664797 (s)
Testing performance with the cache capacity: 1000
Cache put time is 9978186000.000000 (ns) and hit rate is 1000000
Worker time: 4631482000.000000 (ns); Hit rate: 1000000; Started at: 0
Worker time: 4668539000.000000 (ns); Hit rate: 1000000; Started at: 0
Worker time: 4653341000.000000 (ns); Hit rate: 1000000; Started at: 0
Worker time: 4608202000.000000 (ns); Hit rate: 1000000; Started at: 0
Total time: 28.541324 (s)
Testing performance with the cache capacity: 10000
Cache put time is 11768426000.000000 (ns) and hit rate is 1000000
Worker time: 4714246000.000000 (ns); Hit rate: 1000000; Started at: 0
Worker time: 4678632000.000000 (ns); Hit rate: 1000000; Started at: 0
Worker time: 4712865000.000000 (ns); Hit rate: 1000000; Started at: 0
Worker time: 4677736000.000000 (ns); Hit rate: 1000000; Started at: 0
Total time: 30.554077 (s)
Testing performance with the cache capacity: 100000
Cache put time is 10341673000.000000 (ns) and hit rate is 1000000
Worker time: 5460446000.000000 (ns); Hit rate: 1000000; Started at: 0
Worker time: 4904672000.000000 (ns); Hit rate: 1000000; Started at: 0
Worker time: 4909892000.000000 (ns); Hit rate: 1000000; Started at: 0
Worker time: 5552180000.00000 (ns); Hit rate: 1000000; Started at: 0
Total time: 31.170976 (s)
```

### CODE REVIEW

**Comment by kalaiyarasiganeshalingam** [comment]
Closing this PR as we need to check whether can we implement using ballerina map and linked list with java semaphore.

File: cache-native/src/main/java/org/ballerinalang/stdlib/cache/nativeimpl/concurrentlinkedhashmap/Weighers.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
+
```
Comment (ldclakmal) [inline cache-native/src/main/java/org/ballerinalang/stdlib/cache/nativeimpl/concurrentlinkedhashmap/Weighers.java:1]:
```suggestion
```

File: cache-ballerina/cache.bal [hunk 2]
```diff
@@ -108,16 +102,13 @@ public class Cache {
...
-
+        externInit(self);
         decimal? interval = cacheConfig?.cleanupInterval;
         if (interval is decimal) {
             time:Utc currentUtc = time:utcNow();
             time:Utc newTime = time:utcAddSeconds(currentUtc, interval);
             time:Civil time = time:utcToCivil(newTime);
-            var result = task:scheduleJobRecurByFrequency(new Cleanup(self, self.linkedList), interval,
+            var result = task:scheduleJobRecurByFrequency(new Cleanup(self), interval,
                                         startTime = time);
```
Comment (ldclakmal) [inline cache-ballerina/cache.bal:112]:
```suggestion
            var result = task:scheduleJobRecurByFrequency(new Cleanup(self), interval, startTime = time);
```

File: cache-ballerina/cache.bal [hunk 3]
```diff
@@ -141,10 +132,7 @@ public class Cache {
         if (value is ()) {
             return prepareError("Unsupported cache value '()' for the key: " + key + ".");
         }
-        // If the current cache is full (i.e. size = capacity), evict cache.
-        if (self.size() == self.maxCapacity) {
-            evict(self, self.maxCapacity, self.evictionFactor, self.linkedList);
-        }
```
[the remark this answers is not in the document]
Comment (kalaiyarasiganeshalingam) [reply cache-ballerina/cache.bal:147]: Yes. This is handled in java

File: cache-ballerina/cache.bal [hunk 4]
```diff
@@ -52,48 +51,42 @@ boolean cleanupInProgress = false;
...
+    private final decimal & readonly defaultMaxAge;
 
-    # Called when a new `cache:Cache` object is created.
-    # ```ballerina
-    # cache:Cache cache = new({capacity: 10, evictionFactor: 0.2});
-    # ```
+    # Initializes new `cache:Cache` instance.
     #
     # + cacheConfig - Configurations for the `cache:Cache` object
     public isolated function init(CacheConfig cacheConfig = {}) {
```
Comment (daneshk) [inline cache-ballerina/cache.bal:85]:
```suggestion
    public isolated function init(*CacheConfig cacheConfig) {
```

Comment (kalaiyarasiganeshalingam) [reply cache-ballerina/cache.bal:85]: Yes. If we change to this, config initialization will be change.
`Cache cache = new({capacity: 10, evictionFactor: 0.2});` will change to `Cache cache = new(capacity: 10, evictionFactor: 0.2);`

Comment (ldclakmal) [reply cache-ballerina/cache.bal:85]: I tried the following sample. So, seems like this is not. breaking change. @daneshk @kalaiyarasiganeshalingam 
```ballerina
import ballerina/io;

public function main() {
    Bar bar1 = new;
    Bar bar2 = new({ i: 20, s: "Ballerina" });
    Bar bar3 = new(i = 20, s = "Ballerina");
}

public type Foo record {|
    int i = 10;
    string s = "Hello";
|};

public class Bar {
    public function init(*Foo f) {
        io:println(f.i);
        io:println(f.s);
    }
}
```

Comment (kalaiyarasiganeshalingam) [reply cache-ballerina/cache.bal]: Updated in 9209e79

File: cache-ballerina/cache.bal [hunk 5]
```diff
@@ -52,48 +51,42 @@ boolean cleanupInProgress = false;
...
+    private final int & readonly maxCapacity;
+    private final EvictionPolicy & readonly evictionPolicy;
+    private final float & readonly evictionFactor;
+    private final decimal & readonly defaultMaxAge;
 
-    # Called when a new `cache:Cache` object is created.
-    # ```ballerina
-    # cache:Cache cache = new({capacity: 10, evictionFactor: 0.2});
-    # ```
+    # Initializes new `cache:Cache` instance.
```
[the remark this answers is not in the document]
Comment (kalaiyarasiganeshalingam) [reply cache-ballerina/cache.bal:82]: Updated in 9209e79

File: cache-ballerina/cache.bal [hunk 6]
```diff
@@ -52,48 +51,45 @@ boolean cleanupInProgress = false;
...
+    private final int & readonly maxCapacity;
+    private final EvictionPolicy & readonly evictionPolicy;
+    private final float & readonly evictionFactor;
+    private final decimal & readonly defaultMaxAge;
 
-    # Called when a new `cache:Cache` object is created.
+    # Initializes new `cache:Cache` instance.
     # ```ballerina
-    # cache:Cache cache = new({capacity: 10, evictionFactor: 0.2});
+    # cache:Cache cache = new(capacity: 10, evictionFactor: 0.2);
```
Comment (ldclakmal) [inline cache-ballerina/cache.bal:84]:
```suggestion
    # cache:Cache cache = new(capacity = 10, evictionFactor = 0.2);
```

File: cache-ballerina/cache.bal [hunk 7]
```diff
@@ -189,23 +167,13 @@ public class Cache {
...
         time:Utc currentUtc = time:utcNow();
-        if (entry.expTime != -1d && entry.expTime < (<decimal>currentUtc[0] + currentUtc[1])) {
-            self.linkedList.remove(node);
-            externRemove(self, key);
+        any? entry = externGet(self, key, <decimal>currentUtc[0] + currentUtc[1]);
+        if (entry is CacheEntry) {
+            return entry.data;
+        } else {
             return ();
         }
```
Comment (ldclakmal) [inline cache-ballerina/cache.bal:176]:
```suggestion
        return ();
```

File: cache-native/src/main/java/org/ballerinalang/stdlib/cache/nativeimpl/concurrentlinkedhashmap/Weighers.java [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+    /**
+     * A weigher where a value has a weight of <tt>1</tt>. A map bounded with this weigher
+     * will evict when the number of key-value pairs exceeds the capacity.
+     *
+     * @return A weigher where a value takes one unit of capacity.
+     */
+    @SuppressWarnings({
+            "unchecked"
+    })
```
Comment (ldclakmal) [inline cache-native/src/main/java/org/ballerinalang/stdlib/cache/nativeimpl/concurrentlinkedhashmap/Weighers.java:38]:
```suggestion
    @SuppressWarnings("unchecked")
```

**Review by ldclakmal — APPROVED** [review]
LGTM

**Review by daneshk — APPROVED** [review]
LGTM

**Comment by daneshk** [comment]
We need to check code coverage drop

**Comment by kalaiyarasiganeshalingam** [comment]
> We need to check code coverage drop

I will send the new pr for this

**Pull request merged**